### ISSUE TITLE
Bug: [no-unnecessary-condition] value is incorrectly detected when using callbacks

### ISSUE THREAD

**Issue by justasam** [typescript-eslint/typescript-eslint#5376]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=DYUwLgBAhgTjEF4IG0CMAaATOgzAXQChRIBnAewFsQA1KYRCAMzpJAINhgDpGyYBRKAGMAFgAoAdgFcKiAHwQA3gQgQAlowiSZiBElQBKCOSq16SMDCkgA3AQC+B9hq0madI0LITyoLsDIAczEAIjAQEjAQgyA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYZwSgAQBcBPABxQGNoBLEggWhXioDsCB6ZgeztmecXIpkAQ2hE65TswAmVAlSnooiaNE7RIOAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
let arr = [1,2,3]
let someVal = false

arr.forEach(num => {
  if (num === 1) someVal = true;
})

if (someVal) console.log("test")
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

I expected that there would be no error message, as the value can be changed in a `forEach` callback

### Actual Result

There was an error `Unnecessary conditional, value is always falsy. 8:5 - 8:12`

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.30.7` |
| `@typescript-eslint/parser`            | `X.Y.Z` |
| `TypeScript`                           | `4.7.2` |
| `ESLint`                               | `8.15.0` |
| `node`                                 | `X.Y.Z` |

**Comment by JoshuaKGoldberg** [comment]
This is an issue with TypeScript and not typescript-eslint, I'm afraid. It's the same issue linked as [Types Not Correct in/with Callback](https://github.com/microsoft/TypeScript/issues/new?assignees=&labels=Duplicate&template=types-not-correct-in-with-callback.md&title=) under github.com/Microsoft/TypeScript/issues/new. See https://github.com/microsoft/TypeScript/issues/9998 for a full discussion.
